### PR TITLE
Add tokamax megablox kernel

### PR DESCRIPTION
**Pull request by khatwanimohit**
# Description

- adds tokamax as a deps in requirements*.txt files. 
- use ragged_dot api from tokamax hidding behind a flag `use_tokamax_gmm` 

If the change fixes a bug or a Github issue, please include a link, e.g.,:
FIXES: b/436335556

*Notice 1:* Once all tests pass, the "pull ready" label will automatically be assigned.
This label is used for administrative purposes. Please do not add it manually.

*Notice 2:* For external contributions, our settings currently require an approval from a MaxText maintainer to trigger CI tests.

# Tests

Perf and convergence runs done by @suexu1025 in b/436335556#comment11. 
Latest perf run  in https://b.corp.google.com/issues/436335556#comment15

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -37,6 +37,7 @@ tensorflow-datasets
 tensorflow-text
 tensorflow
 tiktoken
+tokamax>=0.0.2
```
Comment (suexu1025) [inline requirements.txt]: do we plan to include latest version in the nightly build?

Comment (khatwanimohit) [reply requirements.txt]: our nightly docker image build only supports using jax,jaxlib,libtpu nightly versions. If we want to have a nightly run then we can prepend `pip install git+https://github.com/openxla/tokamax.git` before the train command which should work

File: src/MaxText/configs/base.yml [hunk 2]
```diff
@@ -860,3 +860,6 @@ subslice_shape: ""
 # NNX
 enable_nnx: false
 shard_optimizer_over_data: False
+
+# Use tokamax library for kernel implementations
+use_tokamax: false
```
Comment (rdyro) [inline src/MaxText/configs/base.yml]: It might be good to distinguish specific tokamax kernels, `use_tokamax_gmm` `use_tokamax_splash` etc

**Review by rdyro — APPROVED** [review]

File: src/MaxText/configs/base.yml (hunk 2 above)
Comment (khatwanimohit) [reply src/MaxText/configs/base.yml]: done

File: src/MaxText/layers/moe.py [hunk 3]
```diff
@@ -807,16 +809,26 @@ def gmm(inputs, kernel, tiling, group_sizes, expert_assignments):
...
-            rhs=kernel,
-            group_sizes=group_sizes,
-            preferred_element_type=self.dtype,
-            tiling=tiling,
-            lhs_quantize_dtype=lhs_quantize_dtype,
-            rhs_quantize_dtype=rhs_quantize_dtype,
-            use_qwix_quantization=self.config.use_qwix_quantization,
-        )
+        if self.config.use_tokamax_gmm:
+          output = api.ragged_dot(
```
Comment (gobbleturk) [inline src/MaxText/layers/moe.py]: can we alias api as tokamax_api or similar?

**Review by gobbleturk — APPROVED** [review]

File: src/MaxText/layers/moe.py [hunk 4]
```diff
@@ -807,16 +809,26 @@ def gmm(inputs, kernel, tiling, group_sizes, expert_assignments):
...
+              lhs=inputs,
+              rhs=kernel,
+              group_sizes=group_sizes,
+              precision=jax.lax.Precision.DEFAULT,
+              preferred_element_type=self.dtype,
+              implementation="mosaic",
```
Comment (NuojCheng) [inline src/MaxText/layers/moe.py:820]: mosaic is TPU-specific, should we disable use_tokamax_gmm for gpu devices?

File: src/MaxText/layers/moe.py (hunk 3 above)
Comment (khatwanimohit) [reply src/MaxText/layers/moe.py]: done

File: src/MaxText/layers/moe.py (hunk 4 above)
Comment (khatwanimohit) [reply src/MaxText/layers/moe.py:820]: nice catch, I'll add a check. I'm also inclined to changing the implementation based on hardware selected.

File: maxtext_jax_ai_image.Dockerfile [hunk 5]
```diff
@@ -51,6 +51,8 @@ RUN if [ "$DEVICE" = "tpu" ] && [ "$JAX_STABLE_STACK_BASEIMAGE" = "us-docker.pkg
 RUN if [ "$DEVICE" = "tpu" ]; then \
         python3 -m pip install 'google-tunix>=0.1.2'; \
         # TODO: Once tunix stopped pinning jax 0.7.1, we should remove our 0.7.0 version pin (b/450286600)
+        # remove tokamax pin once we upgrade to jax 0.8.0
+        python3 -m pip install 'tokamax==0.0.2'; \
```
Comment (bvandermoon) [inline maxtext_jax_ai_image.Dockerfile]: Why is this here if `requirements_with_jax_ai_image.txt` has it set to 0.0.3?

Comment (khatwanimohit) [reply maxtext_jax_ai_image.Dockerfile]: In general I want tokamax installation to be `>=0.0.3`, this pinning only exists as a workaround because pinning to jax=0.7.0 is a workaround. 

once we remove the workarounds and this part of code, I want to installations to happen >= 0.0.3

**Review by bvandermoon — APPROVED** [review]
